File: internal/observability/contracts.go
package observability

import "time"

const (
	OverviewSchema    = "fanout.overview.result@1"
	TopologySchema    = "fanout.topology.result@1"
	PerformanceSchema = "fanout.performance.result@1"
	TraceSchema       = "fanout.trace.result@1"
	LogsSchema        = "fanout.logs.result@1"
)

// Scope is the mandatory boundary for every telemetry query. Callers derive
// Namespace from authenticated application state; models never choose it.
type Scope struct {
	Namespace string    `json:"namespace"`
	Start     time.Time `json:"start"`
	End       time.Time `json:"end"`
}

// Provenance makes every rendered answer traceable to its source query.
type Provenance struct {
	QueryID    string    `json:"query_id"`
	Window     string    `json:"window"`
	Generated  time.Time `json:"generated_at"`
	Complete   bool      `json:"complete"`
	DataSource string    `json:"data_source"`
}

// Result is the canonical query envelope shared by REST and MCP.
// Summary is deliberately compact model context; Data remains authoritative.
// UI selection is deliberately absent: MCP Apps link tools to UI resources
// through standard tool metadata rather than a Fanout-specific view registry.
type Result[T any] struct {
	Schema     string     `json:"schema"`
	Summary    string     `json:"summary"`
	Data       T          `json:"data"`
	Provenance Provenance `json:"provenance"`
}

type Health string

const (
	HealthHealthy   Health = "healthy"
	HealthDegraded  Health = "degraded"
	HealthUnhealthy Health = "unhealthy"
)

type ServiceHealth struct {
	Service     string  `json:"service"`
	Health      Health  `json:"health"`
	Spans       int64   `json:"spans"`
	ErrorRate   float64 `json:"error_rate"`
	P50MS       float64 `json:"p50_ms"`
	P95MS       float64 `json:"p95_ms"`
	LogCount    int64   `json:"log_count"`
	MetricCount int64   `json:"metric_count"`
}

type HealthCounts struct {
	Healthy   int `json:"healthy"`
	Degraded  int `json:"degraded"`
	Unhealthy int `json:"unhealthy"`
}

type Overview struct {
	Health       Health          `json:"health"`
	Counts       HealthCounts    `json:"counts"`
	TotalSpans   int64           `json:"total_spans"`
	ErrorRate    float64         `json:"error_rate"`
	Services     []ServiceHealth `json:"services"`
	ServiceCount int             `json:"service_count"`
}

type Edge struct {
	Caller    string  `json:"caller"`
	Callee    string  `json:"callee"`
	Type      string  `json:"type"`
	Calls     int64   `json:"calls"`
	AverageMS float64 `json:"average_ms"`
	ErrorRate float64 `json:"error_rate"`
}

type Topology struct {
	Nodes []ServiceHealth `json:"nodes"`
	Edges []Edge          `json:"edges"`
}

type PerformancePoint struct {
	Time        time.Time `json:"time"`
	Spans       int64     `json:"spans"`
	ErrorRate   float64   `json:"error_rate"`
	P50MS       float64   `json:"p50_ms"`
	P95MS       float64   `json:"p95_ms"`
	LogCount    int64     `json:"log_count"`
	MetricCount int64     `json:"metric_count"`
}

type Endpoint struct {
	Method    string  `json:"method"`
	Path      string  `json:"path"`
	Calls     int64   `json:"calls"`
	P50MS     float64 `json:"p50_ms"`
	P95MS     float64 `json:"p95_ms"`
	P99MS     float64 `json:"p99_ms"`
	ErrorRate float64 `json:"error_rate"`
	Health    Health  `json:"health"`
}

type HeatmapPoint struct {
	Time    time.Time `json:"time"`
	Service string    `json:"service"`
	P95MS   float64   `json:"p95_ms"`
}

type ComparisonMetric struct {
	Label       string  `json:"label"`
	Unit        string  `json:"unit"`
	Before      float64 `json:"before"`
	After       float64 `json:"after"`
	ChangePct   float64 `json:"change_pct"`
	Direction   string  `json:"direction"`
	Significant bool    `json:"significant"`
}

type Performance struct {
	Service    string             `json:"service,omitempty"`
	Points     []PerformancePoint `json:"points"`
	Endpoints  []Endpoint         `json:"endpoints"`
	Heatmap    []HeatmapPoint     `json:"heatmap"`
	Comparison []ComparisonMetric `json:"comparison"`
}

type TraceSpan struct {
	SpanID        string    `json:"span_id"`
	ParentSpanID  string    `json:"parent_span_id,omitempty"`
	Service       string    `json:"service"`
	Operation     string    `json:"operation"`
	Kind          string    `json:"kind"`
	Start         time.Time `json:"start"`
	DurationMS    float64   `json:"duration_ms"`
	Status        string    `json:"status"`
	StatusMessage string    `json:"status_message,omitempty"`
}

type LogEntry struct {
	Time     time.Time `json:"time"`
	Severity string    `json:"severity"`
	Service  string    `json:"service"`
	Body     string    `json:"body"`
	TraceID  string    `json:"trace_id,omitempty"`
	SpanID   string    `json:"span_id,omitempty"`
}

type TraceDetail struct {
	TraceID    string      `json:"trace_id"`
	DurationMS float64     `json:"duration_ms"`
	HasError   bool        `json:"has_error"`
	Services   []string    `json:"services"`
	Spans      []TraceSpan `json:"spans"`
	Logs       []LogEntry  `json:"logs"`
}

type LogBucket struct {
	Time     time.Time `json:"time"`
	Severity string    `json:"severity"`
	Count    int64     `json:"count"`
}

type Logs struct {
	Entries []LogEntry  `json:"entries"`
	Buckets []LogBucket `json:"buckets"`
}

File: ui/apps/src/contracts.ts
export type Health = "healthy" | "degraded" | "unhealthy";

export interface Provenance {
  query_id: string;
  window: string;
  generated_at: string;
  complete: boolean;
  data_source: string;
}

export interface Result<T> {
  schema: string;
  summary: string;
  data: T;
  provenance: Provenance;
}

export interface ServiceHealth {
  service: string;
  health: Health;
  spans: number;
  error_rate: number;
  p50_ms: number;
  p95_ms: number;
  log_count: number;
  metric_count: number;
}

export interface Overview {
  health: Health;
  counts: { healthy: number; degraded: number; unhealthy: number };
  total_spans: number;
  error_rate: number;
  services: ServiceHealth[];
  service_count: number;
}

export interface Edge {
  caller: string;
  callee: string;
  type: string;
  calls: number;
  average_ms: number;
  error_rate: number;
}

export interface Topology {
  nodes: ServiceHealth[];
  edges: Edge[];
}

export interface PerformancePoint {
  time: string;
  spans: number;
  error_rate: number;
  p50_ms: number;
  p95_ms: number;
  log_count: number;
  metric_count: number;
}

export interface Endpoint {
  method: string;
  path: string;
  calls: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  error_rate: number;
  health: Health;
}

export interface HeatmapPoint {
  time: string;
  service: string;
  p95_ms: number;
}

export interface ComparisonMetric {
  label: string;
  unit: string;
  before: number;
  after: number;
  change_pct: number;
  direction: "improvement" | "regression" | "stable";
  significant: boolean;
}

export interface Performance {
  service?: string;
  points: PerformancePoint[];
  endpoints: Endpoint[];
  heatmap: HeatmapPoint[];
  comparison: ComparisonMetric[];
}

export interface TraceSpan {
  span_id: string;
  parent_span_id?: string;
  service: string;
  operation: string;
  kind: string;
  start: string;
  duration_ms: number;
  status: string;
  status_message?: string;
}

export interface LogEntry {
  time: string;
  severity: string;
  service: string;
  body: string;
  trace_id?: string;
  span_id?: string;
}

export interface TraceDetail {
  trace_id: string;
  duration_ms: number;
  has_error: boolean;
  services: string[];
  spans: TraceSpan[];
  logs: LogEntry[];
}

export interface LogBucket {
  time: string;
  severity: string;
  count: number;
}

export interface Logs {
  entries: LogEntry[];
  buckets: LogBucket[];
}
